src/Agda2Hs/Compile/Type.hs

```diff
@@ -155,42 +107,67 @@ compileType t = do
          | Just semantics <- isSpecialType f -> setCurrentRange f $ semantics es
          | Just args <- allApplyElims es ->
              ifJustM (isUnboxRecord f) (\_ -> compileUnboxType f args) $
-             ifM (isTransparentFunction f) (compileTransparentType args) $
+             ifM (isTransparentFunction f) (compileTransparentType (defType def) args) $
              ifM (isInlinedFunction f) (compileInlineType f es) $ do
-               vs <- compileTypeArgs args
+               vs <- compileTypeArgs (defType def) args
                f <- compileQName f
                return $ tApp (Hs.TyCon () f) vs
          | otherwise -> fail
 
     Var x es | Just args <- allApplyElims es -> do
-      vs <- compileTypeArgs args
+      xi <- lookupBV x
+      unless (usableModality xi) $ genericDocError
+        =<< text "Cannot use erased variable" <+> prettyTCM (var x)
+        <+> text "in Haskell type"
+      vs <- compileTypeArgs (snd $ unDom xi) args
       x  <- hsName <$> compileDBVar x
```

Comment:
`compileDBVar` is used in other places where it's not necessary to check the modality, so I decided to move it here. Since it is just changing where the check happens, I don't think this should cause any breaking change, but if you have a counterexample I'm happy to investigate.